### PR TITLE
PUBDEV-5784 - Add more tests of training cross-validated XGBoost models

### PR DESCRIPTION
**Pull request by Pscheidl**
PR against master to match latest effort :) Includes changes from https://github.com/h2oai/h2o-3/pull/2672

### CODE REVIEW

File: h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/XGBoostTest.java [hunk 1]
```diff
@@ -828,6 +828,248 @@ private Vec createRandomBinaryWeightsVec(final long len, final int randomSeed) {
...
+   * @param len        Length of the resulting fold column vector
+   * @param randomSeed Seed for the random generator (for reproducibility)
+   * @param nfolds     Number of folds. Basically number of categories in the fold column.
+   * @return An instance of {@link Vec} with integers ranging from 0 to nfoldsm representing the number of categories.
+   */
+  private Vec createRandomFoldColumn(final long len, final int nfolds, final int randomSeed) {
+    if (nfolds < 2) throw new IllegalArgumentException("Number of folds must be greater than or equal to 2.");
+    final Vec weightsVec = Vec.makeZero(len, Vec.T_NUM);
+    final Random random = new Random(randomSeed);
+    for (int i = 0; i < weightsVec.length(); i++) {
```
Comment (michalkurka) [inline h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/XGBoostTest.java]: It doesn't matter in the test but we should perpetuate the use of the right APIs in tests as well, please use Vec Writer:

      try (Vec.Writer vw = vec.open()) {
        // for loop with set calls on vw
      }

File: h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/XGBoostTest.java [hunk 2]
```diff
@@ -828,6 +828,248 @@ private Vec createRandomBinaryWeightsVec(final long len, final int randomSeed) {
...
+      assertEquals(foldColumnName, model._output.foldName());
+      parms._fold_column = null; // No longer required
+
+      Key[] originalCrossValidationModels = model._output._cross_validation_models;
+      assertEquals(2, originalCrossValidationModels.length);
+
+      for (int i = 0; i < originalCrossValidationModels.length; i++) {
+
+        crossValidationSubset = Rapids.exec(String.format("(rows %s ( == (cols %s [5]) %d))", iristFrame._key, iristFrame._key, originalCrossValidationModels.length - (i + 1))).getFrame();
+        crossValidationSubset = crossValidationSubset.deepCopy("trainingFrameSubset");
```
Comment (michalkurka) [inline h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/XGBoostTest.java]: Why deepCopy?

File: h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/XGBoostTest.java [hunk 3]
```diff
@@ -828,6 +828,248 @@ private Vec createRandomBinaryWeightsVec(final long len, final int randomSeed) {
...
+      assertEquals(ModelCategory.Regression, model._output.getModelCategory());
+      assertEquals(foldColumnName, model._output.foldName());
+      parms._fold_column = null; // No longer required
+
+      Key[] originalCrossValidationModels = model._output._cross_validation_models;
+      assertEquals(2, originalCrossValidationModels.length);
+
+      for (int i = 0; i < originalCrossValidationModels.length; i++) {
+
+        crossValidationSubset = Rapids.exec(String.format("(rows %s ( == (cols %s [9]) %d))", prostateFrame._key, prostateFrame._key, originalCrossValidationModels.length - (i + 1))).getFrame();
```
Comment (michalkurka) [inline h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/XGBoostTest.java]: each subset should have ideally a unique key and should be deleted explicitly

Comment (Pscheidl) [reply h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/XGBoostTest.java]: It is already deleted explicitely in the finally block. The key is now however generated randomly (that is what I changed).

File: h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/XGBoostTest.java (hunk 2 above)
Comment (Pscheidl) [reply h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/XGBoostTest.java]: Removed. That is a leftover.

**Comment by Pscheidl** [comment]
https://0xdata.atlassian.net/browse/PUBDEV-5784

**Pull request closed without merging**